### ISSUE TITLE
feat(theme-generator): Add possibility to force custom color @angular/material:theme-color

### ISSUE THREAD

**Issue by zygarios** [angular/components#30167]
### Feature Description

My problem is that I still can't understand how `@angular/material:theme-color` works. In the terminal, I need to provide colors for primary, secondary, and tertiary. For each one, I specify a single color, like #ff9500, because that's what I want.

`What HEX color should be used to generate the M3 theme? It will represent your primary color palette. (ex. #ffffff) **#005dba**
What HEX color should be used represent the secondary color palette? (Leave blank to use generated colors from Material) **#005dba**
What HEX color should be used represent the tertiary color palette? (Leave blank to use generated colors from Material) **#005dba**`

 However, when generating the theme, I see that it creates color palettes, but when I try to get just my color, I use **`--mat-sys-primary`.** Unfortunately, it appears significantly darker (#8c5000), and I don't know which variable the color is actually hidden under. I have the impression that due to the low contrast, it is being artificially darkened. However, I selected the option not to generate mixins with high contrast. I would like to be able to use exactly that color without worrying about contrast. Currently, I am forced to manually change all the system token colors because the generated palette still interferes too much with what I want to achieve, taking away the flexibility I need in this case.

### Use Case

It would be great to be able to specify in the terminal that I want exactly the colors I mention, without any interference.

**Comment by jimivdw** [comment]
Very big +1. The fact that you cannot provide exact colors to an M3 theme is a deal breaker for me. Not being able to use my company's brand colors just because they aren't "up to Material's standards" just feels wrong.

I know I can manually override every color variable, and I will do so when support for M2 gets dropped eventually, but this is just very annoying and I really hope there will be an easier way to use exact colors for a theme.

**Comment by amysorto** [comment]
@zygarios For each color specified, those colors are used directly to create color palettes of tones that match that color's hue and chroma. If you only specify a single color or a subset, the other colors get created in relation to each other. So there shouldn't be much adjusting happening if you specify everything and aren't requesting high contrast.

Although, Material defines `--mat-sys-primary` as tone 40 in your primary palette. Typically the color used to create the color palette usually closest to tone 50 in the color palette, which is why it is slightly off to what you are expecting.

Although I do see what you are saying, it would be ideal if you can specify maybe the `primary` color and then the color palette gets extracted from that to get more color accuracy and be more intuitive. I'll leave this issue open so this can be looked into.

In the meantime, overriding colors is your best option for complete customization: https://material.angular.io/guide/theming#system-tokens

**Comment by richardsengers** [comment]
This is a very blocking issue right now. 
It's beautifull having all kinds of colors, but our customers want at least a toolbar with their brand color.

An example (a very ugly primary color but just to let you see the difference):
<img width="185" alt="Image" src="https://github.com/user-attachments/assets/c80afbcf-00e1-46da-868a-29e5077b7702" />

in the new M3
<img width="181" alt="Image" src="https://github.com/user-attachments/assets/c0e76e1f-258f-4ef5-9e95-21a13461cac9" />

**Comment by richardsengers** [comment]
For those interested, I've created a small function which sets all theme tones to css vars, which you can use in your own color palette

```
import { argbFromHex, hexFromArgb, themeFromSourceColor } from '@material/material-color-utilities';

const hueTones = [0, 10, 20, 25, 30, 35, 40, 50, 60, 70, 80, 90, 95, 98, 99, 100];
// Map of neutral hues to the previous/next hues that
// can be used to estimate them, in case they're missing.
const neutralHues = new Map<number, { prev: number; next: number }>([
  [4, { prev: 0, next: 10 }],
  [6, { prev: 0, next: 10 }],
  [12, { prev: 10, next: 20 }],
  [17, { prev: 10, next: 20 }],
  [22, { prev: 20, next: 25 }],
  [24, { prev: 20, next: 25 }],
  [87, { prev: 80, next: 90 }],
  [92, { prev: 90, next: 95 }],
  [94, { prev: 90, next: 95 }],
  [96, { prev: 95, next: 98 }],
]);

const neutralHueTones = [...hueTones, ...neutralHues.keys()];
export function setCSSVariablesFromM3Theme(hex: string) {
  const m3ThemeColorsJSON = themeFromSourceColor(argbFromHex(hex));
  for (const [variant, palette] of Object.entries(m3ThemeColorsJSON.palettes)) {
    const paletteKey = variant.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();
    const tones = paletteKey === 'neutral' ? neutralHueTones : hueTones;
    for (const tone of tones) {
      const color = tone === 40 && paletteKey === 'primary' ? hex : hexFromArgb(palette.tone(tone));
      document.documentElement.style.setProperty(`--m3-${paletteKey}-color-${tone}`, color);
    }
  }
}
```

First you need to `pnpm i @material/material-color-utilities ` which is used internally by the theme-color scheme.

My code is just a little change from what the scheme does
node_modules/@angular/material/schematics/ng-generate/theme-color/index_bundled.js (getColorPalettesSCSS(colorPalettes) function)

Offcourse not an ideal thing to do, but now I can set a brand color to tone 40 and I have all tones as css variables from one hex value.

**Comment by amysorto** [comment]
As a status update, I looked into this a bit further and it isn’t so so straightforward unfortunately. As an example, take these two colors: `#003846` (the darker blue) and `#2ED5FF` (the lighter blue).

<img width="583" alt="Image" src="https://github.com/user-attachments/assets/8dc225b7-dfb0-4e34-9307-46d3043ca23b" />

<img width="582" alt="Image" src="https://github.com/user-attachments/assets/608e1801-6716-403f-a09c-a907aacdb409" />

<img width="881" alt="Image" src="https://github.com/user-attachments/assets/7e9d357f-f5c3-4ccb-bcc4-2ecb77a4200c" />

Tonal palettes are created from a color’s hue, but its tone is adjusted (which are the values seen in the tonal palette image above). Both of these colors share the same hue so they share the same tonal palette. Both of these themes would create the same value for the `primary` color role. The material components theming system picks the color from the palettes for any color role (ex. tone 40 for the primary color role).

The issue with changing the tonal palette itself is how the different colors relate to each other. For example if we tweak the tonal palette so that tone 40 matches either the lighter blue color `#2ED5FF` (which has a tone of ~79), the other tones need to be adjusted to account for color contrast. Although this could get weird because other palettes would also need to be adjusted, but those colors can also be customized. 

It feels like because of the constraints of the material color system, it makes sense for applications who want exact colors to set things themselves to ensure there is adequate color contrast between their palettes. The color roles should be adjusted, not the tonal palettes themselves.

Although having access to the CSS vars of the tonal palettes as mentioned in #30070 is useful so you can do something like this in your application:

```
@include mat.theme-overrides((
  primary: var(—mat-sys-primary80),
));
```

I’ll leave this issue open since it seems useful to provide a more accurate theme to the provided colors. It would likely require more careful planning since it would cause us to move away from Material’s color system to make this work.

**Comment by richardsengers** [comment]
" It would likely require more careful planning since it would cause us to move away from Material’s color system to make this work."

True, but Material 3 theming seems primarily tailored for mobile and Android apps. In contrast, many Angular Material components are used in large web-based applications (based on nothing, just what I think :-)), where achieving the same aesthetic appeal as on Android, particularly with different tonal palette colors, can be more challenging. For instance, consider a dashboard: many applications built with Angular Material include a header featuring a brand color.

Offcourse we can override everything... but like you said it  seems useful to provide a more accurate theme to the provided colors for Angular Material

**Comment by zygarios** [comment]
I have to say that I started using @angular/material:theme-color in a new project with the option to generate tokens as CSS instead of SASS, and I think it's infinitely better. For the first time, I feel like I have proper control over the components, I know what options are available, and I understand where specific properties come from. It's great that some edge-case properties are described in comments. It could be even more detailed, though, because, for example, I still don’t understand the purpose of "fixed" tokens or many from the Neutral group—where they might be needed and whether I can safely remove them. Nonetheless, I really want to appreciate the introduction of CSS-based tokens. Great job, it’s a so much better than mixins with scss! Thank you Angular Material Team😊

**Comment by richardsengers** [comment]
@amysorto it seems like M3 has an option for Color Match

<img width="305" alt="Image" src="https://github.com/user-attachments/assets/820cb4ae-bf36-47d4-9240-1dc6aa51d6c0" />

https://material-foundation.github.io/material-theme-builder/


I think something like this should be added to the scheme, don't you think?
I'm unable to determine what the option actually does behind the scenes.

**Comment by depyronick** [comment]
I've tried to use the generated palette from the material theme builder (the json export) but when I enable the color match option, the sass-loader of angular complains about invalid colors.

I think it would be good to have the option to color match when creating the custom theme with @angular/material:theme-color.

**Comment by zygarios** [comment]
I can highly recommend you to use CSS material generator pallete. Its do much better to use

**Comment by Glavo** [comment]
> [@amysorto](https://github.com/amysorto) it seems like M3 has an option for Color Match
> 
> <img alt="Image" width="305" src="https://private-user-images.githubusercontent.com/810038/407300501-820cb4ae-bf36-47d4-9240-1dc6aa51d6c0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIwOTYzMzcsIm5iZiI6MTc0MjA5NjAzNywicGF0aCI6Ii84MTAwMzgvNDA3MzAwNTAxLTgyMGNiNGFlLWJmMzYtNDdkNC05MjQwLTFkYzZhYTUxZDZjMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxNlQwMzMzNTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03Y2I5NDI5MDI3ODVhYjkzNDY4ZTJiNDc2ODAyODdjODg2YmI1MDdlYjdiYzQ1MzJkM2Y3NWZlZTk1NjA0NTBlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.PT_Hx0lhySxOjhdw61Xmfbw6lnIcn9mxAPo5DeYfJus">
> https://material-foundation.github.io/material-theme-builder/
> 
> I think something like this should be added to the scheme, don't you think? I'm unable to determine what the option actually does behind the scenes.

@richardsengers What this option actually does is set [`dynamicSchemeVariant`](https://api.flutter.dev/flutter/material/DynamicSchemeVariant.html) to `DynamicSchemeVariant.content`.

> Tokens and palettes match the seed color. `ColorScheme.primaryContainer` is the seed color, adjusted to ensure contrast with surfaces. The tertiary palette is analogue of the seed color.

**Comment by peteqian** [comment]
I don't really quite why the the generated palette changes primary color from the one provided in the CLI.

It would be nice to have the additional option to force the color match, since the color I provide is what I want the palette `--mat-sys-primary` to be and not something darker.

**Comment by mattiLeBlanc** [comment]
So I generate my colour palette (primary based on  a colour, and that specific could will not be present in the generated list. That is not very handy.
So then I add my specific color in one of the hue values, it is  hue 80 for this color palette.
But in my theme:
```
 @include mat.theme(
    (
      color: colors.$primary-palette,
      typography: (
        brand-family: Poppins,
        plain-family: 'Inter Tight'
      ),
      density: 0,
    )
  );

```

I set the primary palette and now my but on which is using `--mat-sys-primary` is selecting hue 40.
So now I have to go override this in my button and other components??
This is not making it very easy. In m2 you could define which Hue to use as primary in your palette. 
Why did this functionality get removed?

**Comment by blogcraft** [comment]
All My Clients are raging about this theme change... Sorry but this was the worst update ever. I wish the color pallete to be as it was on M2

**Comment by ankurlorecs** [comment]
This is simply frustrating. Did no one though about enterprise applications which has it's own design system and colors and would not allow to change colors? 

I have this color as primary color.  This is common color. Even metronic theme uses this.
 
![Image](https://github.com/user-attachments/assets/7cf7f193-e728-4cc9-b4a8-54b2c7498f12)

When I use this color as source then below is how my primary color looks like.

![Image](https://github.com/user-attachments/assets/1a06b5f5-0727-41ad-a4f2-154fd51fa92c)

This means that we can't use light blue as primary color at all. What kind of logic is this?

Should we just ask our produce team to redesign their colors to use use m3 specifics? Does this even make sense?

This means that we now have to stick to old version and can't upgrade angular.

This seems to be treated as minor issue. But it's not. This is really really big major issue. This is blocker for existing applications.

This will force large scale applications to move away from angular material and even angular.

Please take this issue seriously.  CC: @amysorto

**Comment by mattiLeBlanc** [comment]
@ankurlorecs I found you just ignore the colour palettes that are generated via the mat-theme-override:

```
  @include mat.theme-overrides(
    (
      background: #07060d,
      on-surface: #fff,
      inverse-on-surface: #000,
      // outline: #fff,
      primary: colors.get-palette-color(colors.$primary-palette, 50),
      on-primary: #fff,
      error: colors.get-palette-color(colors.$error-palette, 50),
      outline: #30578 ;

```

I even corrected some of the colors in the generated palette to match the colour I want. You can see in Angular material which css variable value is used and just override that one.
I agree, the standard generate palettes are useless, they never match the colour our designers intent us to use.

**Comment by ankurlorecs** [comment]
@mattiLeBlanc Yes. But If we just overrides the colors then it defeats the purpose of color pallets at all.
I think this issue should not be labeled as minor issue(p4) and which would not be worked on for long time and becomes stale.

**Comment by mattiLeBlanc** [comment]
@ankurlorecs yeah fair point; what is the usefulness of the generated palette if they do not cover or represent the colours of our company styleguide :/

**Comment by St1c** [comment]
This is extremely annoying issues, since it is not obvious from the docs that using your custom defined colors, doesn't mean, that those will actually be used. [Material builder](https://material-foundation.github.io/material-theme-builder/?primary=%23D7FF93&secondary=%23F3F2EC&tertiary=%236C6C76&neutral=%23DFDBD0&neutralVariant=%23A7A29F&bodyFont=Roboto&displayFont=Roboto&colorMatch=true) actually supports this, but it is not evident how it is actually done...So you really have to dig deep, to understand the issue, and at the end come to this issues thread. 

It seems there is some work going on though...:

https://github.com/andrewseguin/components/commit/c875178bb02924a8427a2e3a8329c7016d01f759

**Comment by lebesnec** [comment]
so to summarize :
- you choose a primary color when generating a palette
- material does not use that color when you use the primary color in your css
- there is no way to access the actual correct color (see https://github.com/angular/components/issues/30070)
